File: app/src/main/java/ch/zuehlke/sbb/reddit/data/source/RedditRepository.java
package ch.zuehlke.sbb.reddit.data.source;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import ch.zuehlke.sbb.reddit.models.RedditNewsData;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by chsc on 08.11.17.
 */

public class RedditRepository implements RedditDataSource {

    private static RedditRepository INSTANCE = null;

    private final RedditDataSource mRedditNewsRemoteDataSource;

    private final RedditDataSource mRedditNewsLocalDataSource;

    /**
     * This variable has package local visibility so it can be accessed from tests.
     */
    Map<String, RedditNewsData> mCacheNews;

    /**
     * Marks the cache as invalid, to force an update the next time redditPost is requested. This variable
     * has package local visibility so it can be accessed from tests.
     */
    boolean mCacheIsDirty = false;


    // Prevent direct instantiation.
    private RedditRepository(@NonNull RedditDataSource newsRemoteDataSource,
                             @NonNull RedditDataSource newsLocalDataSource) {
        mRedditNewsRemoteDataSource = checkNotNull(newsRemoteDataSource);
        mRedditNewsLocalDataSource = checkNotNull(newsLocalDataSource);
    }

    /**
     * Returns the single instance of this class, creating it if necessary.
     *
     * @param newsLocalDataSource the backend redditPost source
     * @param newsLocalDataSource  the device storage redditPost source
     * @return the {@link RedditRepository} instance
     */
    public static RedditRepository getInstance(RedditDataSource newsRemoteDataSource,
                                               RedditDataSource newsLocalDataSource) {
        if (INSTANCE == null) {
            INSTANCE = new RedditRepository(newsRemoteDataSource, newsLocalDataSource);
        }
        return INSTANCE;
    }

    /**
     * Used to force {@link #getInstance(RedditDataSource, RedditDataSource)} to create a new instance
     * next time it's called.
     */
    public static void destroyInstance() {
        INSTANCE = null;
    }



    @Override
    public void getMoreNews(LoadNewsCallback callback) {
        checkNotNull(callback);
        addNewsFromRemoteDataSource(callback);
    }

    private void addNewsFromRemoteDataSource(@NonNull final LoadNewsCallback callback) {
        mRedditNewsRemoteDataSource.getMoreNews(new LoadNewsCallback() {
            @Override
            public void onNewsLoaded(List<RedditNewsData> news) {
                refreshCache(news);
                updateLocalDataSource(news);
                callback.onNewsLoaded(new ArrayList<>(mCacheNews.values()));
            }

            @Override
            public void onDataNotAvailable() {
                callback.onDataNotAvailable();
            }
        });
    }

    @Override
    public void getNews(@NonNull final LoadNewsCallback callback) {
        checkNotNull(callback);
        mRedditNewsRemoteDataSource.refreshNews();
        // Respond immediately with cache if available and not dirty
        if (mCacheNews != null && !mCacheIsDirty) {
            callback.onNewsLoaded(new ArrayList<>(mCacheNews.values()));
            return;
        }

        //TOOD add online check, if not available go local.

        if (mCacheIsDirty) {
            // If the cache is dirty we need to fetch new redditPost from the network.
            getNewsFromRemoteDataSource(new LoadNewsCallback() {
                @Override
                public void onNewsLoaded(List<RedditNewsData> data) {
                    for(RedditNewsData newsData : data){
                        saveRedditNews(newsData);
                    }
                    refreshCache(data);
                    callback.onNewsLoaded(new ArrayList<RedditNewsData>(mCacheNews.values()));
                }

                @Override
                public void onDataNotAvailable() {
                        callback.onDataNotAvailable();
                }
            });
        } else {
            // Query the local storage if available. If not, query the network.
            mRedditNewsLocalDataSource.getNews(new LoadNewsCallback() {
                @Override
                public void onNewsLoaded(List<RedditNewsData> tasks) {
                    refreshCache(tasks);
                    callback.onNewsLoaded(new ArrayList<>(mCacheNews.values()));
                }

                @Override
                public void onDataNotAvailable() {
                    getNewsFromRemoteDataSource(callback);
                }
            });
        }
    }

    @Override
    public void refreshNews() {
        mCacheIsDirty = true;
        mRedditNewsRemoteDataSource.refreshNews();
    }

    @Override
    public void deleteAllNews() {
        mRedditNewsRemoteDataSource.deleteAllNews(); // Although we call deleteAllNews() on the remote datasource, it is not implemented.
        mRedditNewsLocalDataSource.deleteAllNews();

        if (mCacheNews == null) {
            mCacheNews = new LinkedHashMap<>();
        }
        mCacheNews.clear();
    }

    @Override
    public void saveRedditNews(@NonNull RedditNewsData data) {
        checkNotNull(data);

        mRedditNewsLocalDataSource.saveRedditNews(data);
        mRedditNewsRemoteDataSource.saveRedditNews(data); // Although we call saveRedditNews() on the remote datasource, it is not implemented.
        // Do in memory cache update to keep the app UI up to date
        if (mCacheNews == null) {
            mCacheNews = new LinkedHashMap<>();
        }
        mCacheNews.put(data.getId(), data);
    }

    private void getNewsFromRemoteDataSource(@NonNull final LoadNewsCallback callback) {
        mRedditNewsRemoteDataSource.getNews(new LoadNewsCallback() {
            @Override
            public void onNewsLoaded(List<RedditNewsData> news) {
                refreshCache(news);
                refreshLocalDataSource(news);
                callback.onNewsLoaded(new ArrayList<>(mCacheNews.values()));
            }

            @Override
            public void onDataNotAvailable() {
                callback.onDataNotAvailable();
            }
        });
    }

    private void refreshCache(List<RedditNewsData> news) {
        if (mCacheNews == null) {
            mCacheNews = new LinkedHashMap<>();
        }
        mCacheNews.clear();
        for (RedditNewsData data : news) {
            mCacheNews.put(data.getId(), data);
        }
        mCacheIsDirty = false;
    }

    private void updateLocalDataSource(List<RedditNewsData> news){
        for (RedditNewsData data : news) {
            mRedditNewsLocalDataSource.saveRedditNews(data);
        }
    }

    private void refreshLocalDataSource(List<RedditNewsData> news) {
        mRedditNewsLocalDataSource.deleteAllNews();
        for (RedditNewsData data : news) {
             mRedditNewsLocalDataSource.saveRedditNews(data);
        }
    }
}

File: app/src/main/java/ch/zuehlke/sbb/reddit/data/source/RedditRepository.kt
package ch.zuehlke.sbb.reddit.data.source

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo

import java.util.ArrayList
import java.util.LinkedHashMap

import ch.zuehlke.sbb.reddit.models.RedditNewsData
import ch.zuehlke.sbb.reddit.models.RedditPostsData

import com.google.common.base.Preconditions.checkNotNull

/**
 * Created by chsc on 08.11.17.
 */

class RedditRepository// Prevent direct instantiation.
private constructor(newsRemoteDataSource: RedditDataSource,
                    newsLocalDataSource: RedditDataSource, private val mContext: Context) : RedditDataSource {

    private val mRedditNewsRemoteDataSource: RedditDataSource

    private val mRedditNewsLocalDataSource: RedditDataSource

    /**
     * This variable has package local visibility so it can be accessed from tests.
     */
    internal var mCacheNews: MutableMap<String, RedditNewsData>? = null

    /**
     * Marks the cache as invalid, to force an update the next time redditPost is requested. This variable
     * has package local visibility so it can be accessed from tests.
     */
    internal var mCacheIsDirty = false


    init {
        mRedditNewsRemoteDataSource = checkNotNull(newsRemoteDataSource)
        mRedditNewsLocalDataSource = checkNotNull(newsLocalDataSource)
    }


    override fun getMoreNews(callback: RedditDataSource.LoadNewsCallback) {
        checkNotNull(callback)
        addNewsFromRemoteDataSource(callback)
    }

    private fun addNewsFromRemoteDataSource(callback: RedditDataSource.LoadNewsCallback) {
        mRedditNewsRemoteDataSource.getMoreNews(object : RedditDataSource.LoadNewsCallback {
            override fun onNewsLoaded(news: List<RedditNewsData>) {
                refreshCache(news)
                updateLocalDataSource(news)
                callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
            }

            override fun onDataNotAvailable() {
                callback.onDataNotAvailable()
            }
        })
    }

    private val isNetworkAvailable: Boolean
        get() {
            val connectivityManager = mContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetworkInfo = connectivityManager.activeNetworkInfo
            return activeNetworkInfo != null && activeNetworkInfo.isConnected
        }

    override fun getNews(callback: RedditDataSource.LoadNewsCallback) {
        checkNotNull(callback)

        mRedditNewsRemoteDataSource.refreshNews()
        // Respond immediately with cache if available and not dirty
        if (mCacheNews != null && !mCacheIsDirty) {
            callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
            return
        }

        if (!isNetworkAvailable) {
            // Query the local storage if available. If not, query the network.
            mRedditNewsLocalDataSource.getNews(object : RedditDataSource.LoadNewsCallback {
                override fun onNewsLoaded(tasks: List<RedditNewsData>) {
                    refreshCache(tasks)
                    callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
                }

                override fun onDataNotAvailable() {
                    callback.onDataNotAvailable()
                }
            })

        } else {
            if (mCacheIsDirty) {
                // If the cache is dirty we need to fetch new data from the network. The Cache is only dirty, when a refreshNews is going on
                getNewsFromRemoteDataSource(object : RedditDataSource.LoadNewsCallback {
                    override fun onNewsLoaded(data: List<RedditNewsData>) {
                        for (newsData in data) {
                            saveRedditNews(newsData)
                        }
                        refreshCache(data)
                        callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
                    }

                    override fun onDataNotAvailable() {
                        callback.onDataNotAvailable()
                    }
                })
            } else {
                // Query the local storage if available. If not, query the network.
                mRedditNewsLocalDataSource.getNews(object : RedditDataSource.LoadNewsCallback {
                    override fun onNewsLoaded(tasks: List<RedditNewsData>) {
                        refreshCache(tasks)
                        callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
                    }

                    override fun onDataNotAvailable() {
                        getNewsFromRemoteDataSource(callback)
                    }
                })
            }
        }


    }

    override fun getPosts(callback: RedditDataSource.LoadPostsCallback, permalink: String) {
        val convertedPermaLink = convertURLToRemote(permalink)
        mRedditNewsLocalDataSource.getPosts(object : RedditDataSource.LoadPostsCallback {
            override fun onPostsLoaded(posts: List<RedditPostsData>) {
                callback.onPostsLoaded(posts)
            }

            override fun onDataNotAvailable() {

            }
        }, convertedPermaLink)

        mRedditNewsRemoteDataSource.getPosts(object : RedditDataSource.LoadPostsCallback {
            override fun onPostsLoaded(posts: List<RedditPostsData>) {
                mRedditNewsLocalDataSource.deletePostsWithPermaLink(convertedPermaLink)
                for (data in posts) {
                    mRedditNewsLocalDataSource.savePosts(data)
                }
                callback.onPostsLoaded(posts)
            }

            override fun onDataNotAvailable() {

            }
        }, convertedPermaLink)

    }

    private fun convertURLToRemote(url: String): String {
        val parsedUrl = url.substring(url.indexOf(COMMENT_SECION) + COMMENT_SECION.length)
        return parsedUrl.substring(0, parsedUrl.length - 1)
    }

    override fun savePosts(data: RedditPostsData) {

    }

    override fun deletePostsWithPermaLink(permaLink: String) {

    }

    override fun refreshNews() {
        mCacheIsDirty = true
        mRedditNewsRemoteDataSource.refreshNews()
    }

    override fun deleteAllNews() {
        mRedditNewsRemoteDataSource.deleteAllNews() // Although we call deleteAllNews() on the remote datasource, it is not implemented.
        mRedditNewsLocalDataSource.deleteAllNews()

        if (mCacheNews == null) {
            mCacheNews = LinkedHashMap<String, RedditNewsData>()
        }
        mCacheNews!!.clear()
    }

    override fun saveRedditNews(data: RedditNewsData) {
        checkNotNull(data)

        mRedditNewsLocalDataSource.saveRedditNews(data)
        mRedditNewsRemoteDataSource.saveRedditNews(data) // Although we call saveRedditNews() on the remote datasource, it is not implemented.
        // Do in memory cache update to keep the app UI up to date
        if (mCacheNews == null) {
            mCacheNews = LinkedHashMap<String, RedditNewsData>()
        }
        mCacheNews!!.put(data.id, data)
    }

    private fun getNewsFromRemoteDataSource(callback: RedditDataSource.LoadNewsCallback) {
        mRedditNewsRemoteDataSource.getNews(object : RedditDataSource.LoadNewsCallback {
            override fun onNewsLoaded(news: List<RedditNewsData>) {
                refreshCache(news)
                refreshLocalDataSource(news)
                callback.onNewsLoaded(ArrayList(mCacheNews!!.values))
            }

            override fun onDataNotAvailable() {
                callback.onDataNotAvailable()
            }
        })
    }

    private fun refreshCache(news: List<RedditNewsData>) {
        if (mCacheNews == null) {
            mCacheNews = LinkedHashMap<String, RedditNewsData>()
        }
        mCacheNews!!.clear()
        for (data in news) {
            mCacheNews!!.put(data.id, data)
        }
        mCacheIsDirty = false
    }

    private fun updateLocalDataSource(news: List<RedditNewsData>) {
        for (data in news) {
            mRedditNewsLocalDataSource.saveRedditNews(data)
        }
    }

    private fun refreshLocalDataSource(news: List<RedditNewsData>) {
        mRedditNewsLocalDataSource.deleteAllNews()
        for (data in news) {
            mRedditNewsLocalDataSource.saveRedditNews(data)
        }
    }

    companion object {

        private var INSTANCE: RedditRepository? = null

        private val COMMENT_SECION = "comments/"

        /**
         * Returns the single instance of this class, creating it if necessary.

         * @param newsLocalDataSource the backend redditPost source
         * *
         * @param newsLocalDataSource  the device storage redditPost source
         * *
         * @return the [RedditRepository] instance
         */
        fun getInstance(newsRemoteDataSource: RedditDataSource,
                        newsLocalDataSource: RedditDataSource, context: Context): RedditRepository {
            if (INSTANCE == null) {
                INSTANCE = RedditRepository(newsRemoteDataSource, newsLocalDataSource, context)
            }
            return INSTANCE
        }

        /**
         * Used to force [.getInstance] to create a new instance
         * next time it's called.
         */
        fun destroyInstance() {
            INSTANCE = null
        }
    }
}
